timers: fix validation

ShogunPanda opened the pull request:
This PR improves `timers/promises` in the following way:

1. `timers.set*` now use the same validation logic. All arguments, including the `delay` argument (where appropriate) are now validated.
2. `timers.scheduler.wait` now supports the `options.ref` option.
3. Fixed the documentation for `timers.scheduler.wait`. It was not just "roughly" equivalent to `timers.setTimeout`, it was just a wrapper around it. I updated the docs.

H4ad commented:
I like the idea but maybe doing this is an argument to introduce `sleep` on `timers`, a sync version of it (which will probably block the event loop), which can be implemented with `Atomic.wait`.

benjamingr reviewed (dismissed):
I don't think we should do this (roll our third new promises timers API).

Also right below it you have scheduler.wait (4 letters) we can bind it to the scheduler which would let you do `wait` (4 letters) vs. `sleep` (5 letters) if that's a meaningful metric.

ShogunPanda commented:
@benjamingr That's different, IMHO. The scheduler API is experimental and thus might go away at anytime.

`setTimeout` is not going anywhere anytime soon.

The number of letters doesn't really matter (thanks autocompletion ;)) but I'd like to give users a simple API they can use without the risk of overriding a well known global and therefore having side effects.

benjamingr commented:
@ShogunPanda we currently have 4 "official" ways to set timeouts (setTimeout, promises setTimeout, scheduler.postTask and scheduler.wait) as well as lots of others for specific timeouts (e.g. AbortSignal.timeout) . That's... already too many ways to do the same thing anyway, I doubt adding another would improve things.

If there is a consensus I'll remove my block.  

> The scheduler API is experimental and thus might go away at anytime.

So is any new API we add like this one?

ShogunPanda commented:
I agree: let's wait for other consensus and can disambiguate this.

ShogunPanda commented:
@benjamingr @jasnell @legendecas 

As agreed in [TSC/1608](https://github.com/nodejs/TSC/pull/1608) I started working on the warning for `timers.setTimeout`.

I then noticed that it was not the documentation that was lacking but rather the validation.
I therefore refactored and unified the validation for `timers.setTimeout`, `timers.setInterval` and `timers.setImmediate`, ensuring also the delay argument is properly validated.

Also, I noticed that documentation for `scheduler.wait` was incorrect: `scheduler.wait` is totally equivalent (as it internally uses it) to call `timers.setTimeout`. Therefore we can support the `options.ref` as well. I updated code and documentation to reflect that.

At the end of the day, since I just refactored internal parts (`timers.set*`) and add support for `options.ref` in `scheduler.wait` (which is experimental anyway) I think this PR should be semver-minor.

ovflowd reviewed (commented):
So is this pretty much a refactor/cleanup? And the only thing actually being fixed are the docs?

ovflowd reviewed (approved):
It looks clean to me -- LGTM

ShogunPanda commented:
Kinda, yes

aduh95 commented on lib/timers/promises.js:
```diff
@@ -136,13 +131,29 @@ function setImmediate(value, options = kEmptyObject) {
+    if (typeof options?.ref !== 'undefined') {
+      validateBoolean(options.ref, 'options.ref');
+    }
+  } catch (err) {
+    return PromiseReject(err);
+  }
```
Why catch the error in the first place? Isn't returning a promise rejection strictly equivalent in an async function?

ShogunPanda replied:
Yes, I copied and pasted from `setTimeout` (which is not async) and missed that. Do you think we can leave like that or do we need a tiny fixing PR?

aduh95 replied:
It would have been more "clean" to close that PR about adding an alias for `setTimeout`, and open a new one for the validation changes. That way, folks watching the repo would have had more chance to notice the change, and our wait time policy would have give enough time for folks to chime in (all the changes happen during a weekend, making it even more likely fewer folks had time to have a look). All in all, I don't really care if you do the follow-up PR with my suggestion, but IMO it's the "right" thing to do.

ovflowd commented:
I feel this landed prematurely. Although I gave approval I believe having at least an extra one would have been good.

ShogunPanda commented on doc/api/timers.md:
```diff
@@ -532,9 +532,8 @@ added:
-Calling `timersPromises.scheduler.wait(delay, options)` is roughly equivalent
-to calling `timersPromises.setTimeout(delay, undefined, options)` except that
-the `ref` option is not supported.
+Calling `timersPromises.scheduler.wait(delay, options)` is equivalent
+to calling `timersPromises.setTimeout(delay, undefined, options)`.
```
Because the documentation was incorrect. It was not `roughly` equivalent, it was `strictly` equivalent since it just wrapping it. I made that explicit.

aduh95 replied:
Then I guess we should document `ref` as a valid option

ronag reviewed (commented):
LGTM
